src/measure_theory/measure/lebesgue.lean

```diff
@@ -239,11 +261,152 @@ lemma map_volume_mul_right {a : ℝ} (h : a ≠ 0) :
+begin
+  refine (measure.pi_eq (λ s hs, _)).symm,
+  simp only [det_diagonal, measure.coe_smul, algebra.id.smul_eq_mul, pi.smul_apply],
+  rw [measure.map_apply _ (measurable_set.univ_pi_fintype hs)],
+  swap, { exact continuous.measurable (linear_map.continuous_on_pi _) },
+  have : (matrix.to_lin' (diagonal D)) ⁻¹' (set.pi set.univ (λ (i : ι), s i))
+    = set.pi set.univ (λ (i : ι), ((*) (D i)) ⁻¹' (s i)),
+  { ext f,
+    simp only [linear_map.coe_proj, algebra.id.smul_eq_mul, linear_map.smul_apply, mem_univ_pi,
+      mem_preimage, linear_map.pi_apply, diagonal_to_lin'] },
+  have B : ∀ i, of_real (abs (D i)) * volume (has_mul.mul (D i) ⁻¹' s i) = volume (s i),
+  { assume i,
+    have A : D i ≠ 0,
+    { simp only [det_diagonal, ne.def] at h,
+      exact finset.prod_ne_zero_iff.1 h i (finset.mem_univ i) },
+    rw [volume_preimage_mul_left A, ← mul_assoc, ← ennreal.of_real_mul (abs_nonneg _), ← abs_mul,
+      mul_inv_cancel A, abs_one, ennreal.of_real_one, one_mul] },
+  rw [this, volume_pi_pi, finset.abs_prod,
+    ennreal.of_real_prod_of_nonneg (λ i hi, abs_nonneg (D i)), ← finset.prod_mul_distrib],
+  { simp only [B] },
+  { exact λ i, measurable_const_mul _ (hs i) },
+end
+
+/-- A transvection preserves Lebesgue measure. -/
+lemma map_transvection_volume_pi [decidable_eq ι] (t : transvection_struct ι ℝ) :
+  measure.map (t.to_matrix.to_lin') volume = volume :=
+begin
+  /- We separate the coordinate along which there is a shearing from the other ones, and apply
+  Fubini. Along this coordinate (and when all the other coordinates are fixed), it acts like a
+  translation, and therefore preserves Lebesgue. -/
+  suffices H : measure_preserving t.to_matrix.to_lin' volume volume, by exact H.2,
+  let p : ι → Prop := λ i, i ≠ t.i,
+  let α : Type* := {x // p x},
+  let β : Type* := {x // ¬ (p x)},
+  let g : (α → ℝ) → (β → ℝ) → (β → ℝ) := λ a b, (λ x, t.c * a ⟨t.j, t.hij.symm⟩) + b,
+  let F : (α → ℝ) × (β → ℝ) → (α → ℝ) × (β → ℝ) :=
+    λ p, (id p.1, g p.1 p.2),
+  let e := equiv.pi_equiv_pi_subtype_prod p (λ (i : ι), ℝ),
+  have : (t.to_matrix.to_lin' : (ι → ℝ) → (ι → ℝ)) = e.symm ∘ F ∘ e,
+  { cases t,
+    ext f k,
+    simp only [linear_equiv.map_smul, dite_eq_ite, linear_map.id_coe, p, ite_not,
+      algebra.id.smul_eq_mul, one_mul, dot_product, std_basis_matrix,
+      equiv.pi_equiv_pi_subtype_prod_symm_apply, id.def, transvection,
+      pi.add_apply, zero_mul, linear_map.smul_apply, function.comp_app,
+      equiv.pi_equiv_pi_subtype_prod_apply, matrix.transvection_struct.to_matrix_mk,
+      matrix.mul_vec, linear_equiv.map_add, ite_mul, e, matrix.to_lin'_apply,
+      pi.smul_apply, subtype.coe_mk, g, linear_map.add_apply, finset.sum_congr, matrix.to_lin'_one],
+    by_cases h : t_i = k,
+    { simp only [h, true_and, finset.mem_univ, if_true, eq_self_iff_true, finset.sum_ite_eq,
+        one_apply, boole_mul, add_comm], },
+    { simp only [h, ne.symm h, add_zero, if_false, finset.sum_const_zero, false_and, mul_zero] } },
```

Comment:
I tried replicating this, starting from a `simp` without arguments, and adding lemmas to unfold things depending on the result I got, and it was not that easy. I think that we are missing simp lemmas.

It looks like `simp [e, g, F, transvection, matrix.mul_vec, dot_product, std_basis_matrix]` gets to something close to the result of the first `simp only`, but `simp [e, g, F, transvection, matrix.mul_vec]` does not because simp does not know how to simplify `dot_product (std_basis_matrix k t_j t_c k) f`. That's one simp lemma that could be added.

Also, that big have proof could be replaced by this:
```suggestion
    by_cases h : t_i = k,
    { simp [e, g, F, transvection, matrix.mul_vec, dot_product, std_basis_matrix, h, add_comm] },
    { simp [e, g, F, transvection, matrix.mul_vec, dot_product, std_basis_matrix, h, ne.symm h] } },
```

Reply:
To me, the proof you are giving shows that we have enough simp lemmas: just tell Lean to expand a few things (which usually should not be expanded), and then simp is able to complete the proof. The fact that I use `simp only` instead of `simp` is just to speed up the proof: on my computer, your version takes 18.4s while the version with the `simp only`s takes only 2.87s. The difference is big enough that I think the `simp only` are worth it, even though it's less readable. The goal of the first big simp only is to factor out the parts that are common to both proofs, precisely to make it just a little bit more readabale.

Reply:
The factored proof you give is certainly better than two enormous `simp only` for the two cases. And indeed computing the result of the dot product I mentioned may not be always what we want.